builder/digest_remote.go

```diff
@@ -0,0 +1,92 @@
+package builder
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/Azure/acr-builder/graph"
+	"github.com/Azure/acr-builder/pkg/image"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/docker/distribution/reference"
+	"github.com/pkg/errors"
+)
+
+type remoteDigest struct {
+	registryCreds graph.RegistryLoginCredentials
+}
+
+func NewRemoteDigest(creds graph.RegistryLoginCredentials) *remoteDigest {
+	return &remoteDigest{
+		registryCreds: creds,
+	}
+}
+
+var _ DigestHelper = &remoteDigest{}
+
+func (d *remoteDigest) PopulateDigest(ctx context.Context, ref *image.Reference) error {
+	if ref == nil {
+		return nil
+	}
+	if ref.Digest != "" {
+		return nil
+	}
+	client := http.DefaultClient
+	opts := docker.ResolverOptions{
+		Client: client,
+	}
+	if cred, ok := d.registryCreds[ref.Registry]; ok {
+		if cred.Username.ResolvedValue == "" || cred.Password.ResolvedValue == "" {
+			return fmt.Errorf("Error fetching credentials for '%s'", ref.Registry)
+		}
+		// Adds credential resolver if private registry
+		opts.Credentials = func(hostName string) (string, string, error) {
+			return cred.Username.ResolvedValue, cred.Password.ResolvedValue, nil
+		}
+	}
+
+	resolver := docker.NewResolver(opts)
+	imageRef, err := getReferencePath(ref)
+	if err != nil {
+		return err
+	}
+	// Note: (bug), if the final image is NOT pushed in this build,
```

Comment:
@shizhMSFT @northtyphoon please review this note

Reply:
I remember current implementation based on docker inspect will return empty digest if the image is not pushed. I think we need to keep the same behavior.
For the following case, I think we still can use docker inspect to query the digest of `$Registry/acb:lates`. If it's pushed, the code will return the digest. If it's not pushed, the code will return empty.
For base image, the code will switch to manifest api to get the digest for buildkit build.

Reply:
Great idea, thanks.